### PR TITLE
Improved T3T1 framebuffer switching

### PR DESCRIPTION
**Pull request by cepetr**
This PR improves the display refreshing algorithm for the T3T1 model with the following enhancements:

1. Simplified the framebuffer switching algorithm, enabling full framerate, especially for simpler animations.
2. Modified the semantics of `display_get_frame_buffer()` and `display_refresh()`. Details are provided below.
3. Fixed the issue of waiting for the TE signal on the T2T1 model (with `NEW_RENDERING`=1).
4. Shorten `ANIM_FRAME_TIMER` deadline from 18ms to 1ms

The functions `display_get_frame_buffer()` and `display_refresh()` should now be used as a pair:

On the T3T1 model, all framebuffer switching (except when running in handler mode) is done in the background.

- `display_get_frame_buffer()`: Retrieves a new framebuffer for writing. This function may block for a while if the buffer is not ready. If called twice consecutively without an intervening call to `display_refresh()`, it returns the same buffer.
- `display_refresh()`: Enqueues the last retrieved buffer. This function is always non-blocking. Calling `display_refresh()` twice consecutively without an intervening call to `display_get_frame_buffer()` has no effect.

### CODE REVIEW

File: core/embed/trezorhal/stm32u5/bg_copy.c [hunk 1]
```diff
@@ -31,7 +32,7 @@ void HAL_DMA_XferCpltCallback(DMA_HandleTypeDef *hdma) {
   }
 }
 
-void GPDMA1_Channel0_IRQHandler(void) {
+__attribute__((optimize("-O0"))) void GPDMA1_Channel0_IRQHandler(void) {
```
Comment (TychoVrahe) [inline core/embed/trezorhal/stm32u5/bg_copy.c:35]: why O0?

**Review by TychoVrahe — CHANGES_REQUESTED** [review]
Just first test and quick look over the code.

I like the queue idea (though it should probably be moved outisde the st-7789 driver as its independent and can be useful for other displays too)

Testing revealed that this change reintroduce the problem i was solving by waiting for two TE events: sometimes, a frame is not displayed or not displayed completely. So further investigation into why is this happening and how it should be prevented is needed - i still think that in principle the system here should work, but maybe we have something in the driver misconfigured or there is something else i don't understand.

File: core/embed/trezorhal/stm32u5/bg_copy.c (hunk 1 above)
Comment (cepetr) [reply core/embed/trezorhal/stm32u5/bg_copy.c]: It was here just for debugging, I forgot to remove it.. fixed here 294c6d52894e33d6a19dae14f4dd42734faa0c57

**Comment by cepetr** [comment]
> I like the queue idea (though it should probably be moved outisde the st-7789 driver as its independent and can be useful for other displays too)

Yes, I agree. When we implement the display driver for T3W1, we should consider it. Currently, there's only one other driver with two frame buffers - DISC2, which is probably not worth improving.

**Comment by cepetr** [comment]
> Testing revealed that this change reintroduce the problem i was solving by waiting for two TE events: sometimes, a frame is not displayed or not displayed completely. So further investigation into why is this happening and how it should be prevented is needed - i still think that in principle the system here should work, but maybe we have something in the driver misconfigured or there is something else i don't understand.

Found the problem and fixed here  624537322c854c6d0b47f1d3d5e80dc2cade5993.

(When TE interrupt was disabled and we called `display_refresh()`, we were not waiting for rising edge of TE signal properly.)

**Review by TychoVrahe — APPROVED** [review]
Looks great, pointing just a few of typos below

File: core/embed/trezorhal/stm32f4/xdisplay/st-7789/display_fb.c [hunk 2]
```diff
@@ -41,173 +42,219 @@
...
-    memcpy(physical_frame_buffer_0, physical_frame_buffer_1,
-           sizeof(physical_frame_buffer_0));
+  switch (drv->queue.entry[drv->queue.rix]) {
+    case FB_STATE_EMPTY:
+    case FB_STATE_PREPARING:
+      // No new frame queued
+      break;
+
+    case FB_STATE_COPYING:
+      // Currenly we are copying a data to the display.
```
[the remark this answers is not in the document]
Comment (cepetr) [reply core/embed/trezorhal/stm32f4/xdisplay/st-7789/display_fb.c]: fixed, see 3ce832c65920e5dcc6ec419aa52c531503bf23eb

File: core/embed/trezorhal/stm32f4/xdisplay/st-7789/display_fb.c [hunk 3]
```diff
@@ -41,173 +42,219 @@
...
-           sizeof(physical_frame_buffer_1));
+      // NOTE: when copying is done, this queue slot is marked empty
+      // (see bg_copy_callback())
+      break;
 
-    pending_fb_switch = 1;
-    __HAL_GPIO_EXTI_CLEAR_FLAG(DISPLAY_TE_PIN);
-    svc_enableIRQ(DISPLAY_TE_INTERRUPT_NUM);
+    default:
+      // This is invalid state and we should never get here
```
[the remark this answers is not in the document]
Comment (cepetr) [reply core/embed/trezorhal/stm32f4/xdisplay/st-7789/display_fb.c]: fixed, see 3ce832c65920e5dcc6ec419aa52c531503bf23eb

File: core/embed/trezorhal/stm32f4/xdisplay/st-7789/display_fb.c [hunk 4]
```diff
@@ -41,173 +42,219 @@
...
   if (is_mode_handler()) {
-    if (pending_fb_switch != 0) {
-      pending_fb_switch = 0;
-      bg_copy_abort();
+    // Disable scheduling of any new background copying
+    HAL_NVIC_DisableIRQ(DISPLAY_TE_INTERRUPT_NUM);
+    // Wait for next TE signal. During this time the
+    // display might be updated in the background
+    wait_for_te_signal();
+    // Stop any backround copying even if it is not finished yet
```
[the remark this answers is not in the document]
Comment (cepetr) [reply core/embed/trezorhal/stm32f4/xdisplay/st-7789/display_fb.c]: fixed, see 3ce832c65920e5dcc6ec419aa52c531503bf23eb

**Pull request merged**